### DISCUSSION TITLE
Packing own esptool version in the code

### DISCUSSION THREAD

**Discussion by sgiath** [markqvist/Reticulum#532] [General]
Hi, I am trying to use the `rnodeconf --autoinstall` on NixOS and I have bumped into an issue that the program is generating `.config/rnodeconf/update/1.72/esptool.py` file and executing it with system Python, which does not have `pyserial` available on NixOS. We have discussed in nixpkgs repo if we could/should use official `esptool` package instead of generating the code on the client machine. So I wanted to ask what is the reasoning behind the current approach? Is the code somehow patched? Is specific version needed? Is there a reason normal `esptool` package should not be used?

Here is the original issue https://github.com/NixOS/nixpkgs/issues/336766
Here is the PR to patch the code to use the correct Python version on NixOS https://github.com/NixOS/nixpkgs/pull/337285

**Comment by sgiath** [comment]
I have found this comment https://github.com/markqvist/Reticulum/discussions/60#discussioncomment-2839631 - is this the sole reason behind it?

cc @markqvist

**Reply by markqvist** [reply]
That, and the fact that the `esptool` package may not be installed or available on a particular system, and that basic RNS bootstrapping and operating functionalities should work with as few external dependencies as entirely possible.

I'm not very familiar with the internals of NixOS. Do you have any good ideas for how to handle this in a graceful way that would make it work on Nix?

**Reply by sgiath** [reply]
It is solved in NixOS by patching the source code to use the esptool package code from the repository.

Here is the patch file:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/rns/unvendor-esptool.patch

The `@esptool@` is replaced during the build process with `esptool` executable:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/rns/default.nix#L29-L31

**Reply by markqvist** [reply]
That's a good way to go about it.
